app/javascript/crayons/MultiInput/MultiInput.jsx

```diff
@@ -17,90 +17,219 @@ const ALLOWED_CHARS_REGEX = /([a-zA-Z0-9@.])/;
  * @param {string} props.placeholder Input placeholder text
  */
 
-export const MultiInput = ({ placeholder }) => {
+export const MultiInput = ({
+  placeholder,
+  regex,
+  SelectionTemplate = DefaultSelectionTemplate,
+}) => {
   const inputRef = useRef(null);
+  const inputSizerRef = useRef(null);
+  const selectedItemsRef = useRef(null);
+
   const [items, setItems] = useState([]);
+  const [editValue, setEditValue] = useState(null);
+  const [inputPosition, setInputPosition] = useState(null);
+
+  useEffect(() => {
+    // editValue defaults to null when component is first rendered.
+    // This ensures we do not autofocus the input before the user has started interacting with the component.
+    if (editValue === null) {
+      return;
+    }
 
-  const handleBlur = ({ target: { value } }) => {
+    const { current: input } = inputRef;
+    if (input && inputPosition !== null) {
+      // Entering 'edit' mode
+      resizeInputToContentSize();
+      input.value = editValue;
+      const { length: cursorPosition } = editValue;
+      input.focus();
+      // This will set the cursor position at the end of the text.
+      input.setSelectionRange(cursorPosition, cursorPosition);
+    }
+  }, [inputPosition, editValue]);
+
+  const handleInputBlur = ({ target: { value } }) => {
     addItemToList(value);
-    clearSelection();
+    clearInput();
+  };
+
+  const handleInputChange = async ({ target: { value } }) => {
+    // When the input appears inline in "edit" mode, we need to dynamically calculate the width to ensure it occupies the right space
+    // (an input cannot resize based on its text content). We use a hidden <span> to track the size.
+    inputSizerRef.current.innerText = value;
+
+    if (inputPosition !== null) {
+      resizeInputToContentSize();
+    }
   };
 
   const handleKeyDown = (e) => {
+    const { value: currentValue } = inputRef.current;
+
     switch (e.key) {
       case KEYS.SPACE:
       case KEYS.ENTER:
       case KEYS.COMMA:
         e.preventDefault();
         addItemToList(e.target.value);
-        clearSelection();
+        clearInput();
+        break;
+      case KEYS.DELETE:
+        if (currentValue === '') {
+          e.preventDefault();
+          editPreviousSelectionIfExists();
+        }
         break;
       default:
-        if (!ALLOWED_CHARS_REGEX.test(e.key)) {
+        if (!regex.test(e.key)) {
```

Comment:
(nitpick) - I don't think we've noted `regex` as a required prop, but we also don't default its value of line 22 & expect it to exist here 🤔 

Maybe instead we should cover the scenario where no regex is provided and "anything goes":
```suggestion
        if (regex && !regex.test(e.key)) {
```

Reply:
I like this, it makes a lot of sense, thanks for noting this!